### DISCUSSION TITLE
Change of normal direction after delaunay2d

### DISCUSSION THREAD

**Discussion by pkuehne22** [pyvista/pyvista#5306] [Q&A]
Hi everyone, 

I have two diffrent [polydatas ](https://github.com/pyvista/pyvista/files/13614993/polydata_files.zip), which both contain information about ground surfaces. I care about the surface normals of those ground surfaces. When Im plotting the surface normals I can see that in both cases the normals point mostly downwards (the average normal is close to [0, 0, -1]). But after using delaunay2d on both datasets the normals flips in the one case and in the other case not.  One can see this when calculating the average normal too.  Does anyone has an idea why that happens? 

With the following script you can reproduce my problem.

```
import pyvista as pv
import numpy as np

polydata1 = pv.read("ground_surfaces1.vtk")
polydata1 = polydata1.compute_normals(
    cell_normals=True, point_normals=False, consistent_normals=True
)
print("average normal for polydata 1 pre delaunay:", np.average(polydata1.cell_data["Normals"], axis=0))

polydata1 = polydata1.delaunay_2d()
polydata1 = polydata1.compute_normals(
    cell_normals=True, point_normals=False, consistent_normals=True
)
print("average normal for polydata 1 after delaunay:", np.average(polydata1.cell_data["Normals"], axis=0))

# -----------------------------------------------------------------------------

polydata2 = pv.read("ground_surfaces2.vtk")
polydata2 = polydata2.compute_normals(
    cell_normals=True, point_normals=False, consistent_normals=True
)
print("average normal for polydata 2 pre delaunay:", np.average(polydata2.cell_data["Normals"], axis=0))

polydata2 = polydata2.delaunay_2d()
polydata2 = polydata2.compute_normals(
    cell_normals=True, point_normals=False, consistent_normals=True
)
print("average normal for polydata 2 after delaunay:", np.average(polydata2.cell_data["Normals"], axis=0))
```
Greetings Philip

**Comment by darikg** [comment]
`delaunay_2d` calls [SetProjectionPlaneMode](https://vtk.org/doc/nightly/html/classvtkDelaunay2D.html#a9eefaa77ce478b3b02a11cfba1241714) to `VTK_BEST_FITTING_PLANE` [here](https://github.com/pyvista/pyvista/blob/5b0d2ab0a3438a9eb63e921fada24a53f8e48b19/pyvista/core/filters/poly_data.py#L2644)

I think what's happening is that VTK fits a plane to the pointset, ignoring face normals, and the normal of that plane is numerically unstable.

Couple of options:
  - Modify `delaunay_2d` to accept some extra options:
    -   `VTK_DELAUNAY_XY_PLANE` (=0) to not fit a plane, and simply ignore the z-coordinates. This will at least give you consistent normals, although they'll aways be in the positive z-direction
    - `VTK_SET_TRANSFORM_PLANE` (=2) and [set a custom transform](https://vtk.org/doc/nightly/html/classvtkDelaunay2D.html#ac1db109f021718dcb643dc56e33f711d)
  - Just check the average normal like you're already doing and use the 'flip_normals' option in `compute_normals`

**Reply by pkuehne22** [reply]
Thanks for the fast reply. 

Perhaps I expressed myself a bit poorly. I want the result of delaunay2d to be the best fitting plane but when applying delaunay2d to two polydatas which normals are pointing in the same directions the output should be consistent. I don’t want that sometimes the normals are flipped and other times not. Checking the average normal after delaunay2d would work but I hoped for an more elegant solution. 

Greetings Philip
